docs/features/incremental-generators.md

```diff
@@ -0,0 +1,485 @@
+that utilizes LINQ:
+
+```csharp
+    // using System.Linq;
+    var singleOp = additionalTexts.BatchTransform(texts => texts.Select(text => text.Path));
+```
+
+**OPEN QUESTION** Should there be versions of
+`BatchTransform`/`BatchTransformMany` that take no transformation, and just
+perform the identity function as specified above?
+
+### Outputting values
+
+At some point in the pipeline the author will want to actually use the
+transformed data to produce an output, such as a `SourceText`. For this purpose
+there are 'terminating' extension methods that allow the author to provide the
+resulting data the generator produces. For instance there is a
+`GenerateSource(...)` method:
+
+``` csharp
+static partial class IncrementalValueSourceExtensions
+{
+    public internal static IncrementalGeneratorOutput GenerateSource<T>(this IncrementalValueSource<T> source, Action<SourceProductionContext, T> action) => ...
+}
+```
+
+That can be used by the author to supply `SourceText` to be appended to the
+compilation via the passed in `SourceProductionContext`:
+
+```csharp
+    // create an output that takes the file paths from the above batch and makes some user visible syntax
+    var output = batched.GenerateSource((sourceContext, filePaths) =>
+    {
+        sourceContext.AddSource("additionalFiles.cs", @"
+namespace Generated
+{
+    public class AdditionalTextList
+    {
+        public static void PrintTexts()
+        {
+            System.Console.WriteLine(""Additional Texts were: " + string.Join(", ", filePaths) + @" "");
+        }
+    }
+}");
+    });
+
+```
+
+After creation, output nodes must be then registered with the pipeline context:
+
+```csharp
+    context.RegisterOutput(output);
+```
+
+A generator may create and register multiple output nodes as part of the
+pipeline, but an output cannot be further transformed once it is created.
+
+It is expected that there will be equivalent `Generate...` methods for
+generating non source outputs such as `GenerateArtifact(...)` or
+`GenerateEmbeddedFile(...)`. Splitting the outputs by type produced allows the
+host (such as the IDE) to not run outputs that are not required. For instance
+artifacts are only produced as part of a command line build, so the IDE has no
+need to run the artifact based outputs or steps that feed into it.
+
+**OPEN QUESTION:** Should we just make the registration automatic, as part of
+calling `.GenerateSource`? Its trickier to plumb through behind the scenes, but
+I don't see any obvious use case for an author to create a node and not register
+it? I've footgunned myself already with this where I forgot to register the
+output and wondered why nothing was happening.
+
+**OPEN QUESTION** Should we have `GenerateBatch...` versions of the output
+methods? This can already be achieved by the author calling `BatchTransform`
+before calling `Generate...` so would just be a helper, but seems common enough
+that it could be useful.
+
+### Simple example
```

Comment:
Worry a bit that the document takes this long to get to the simple example. Generally I would try to get this in earlier, basically "here is what an incremental generator looks like" then use later sections to break down the concepts involved.

Basically show readers up front this is what a simple example looks like. Put a working model in their head and then introduce the concepts that they can keep mapping back to that initial model. The way it's written it's a lot of concepts that are hard to necessarily map back to a working example.